### PR TITLE
Make return value of Controller:referer() consistent when returning local URLs.

### PR DESCRIPTION
**Pull request by ADmad**
Before this fix if referrer is available the local URL returned does not have
base appended but if a referrer is not found the URL returned using the default
URL passed as argument gets returned with base appended. This inconsistency
makes it difficult to have usage like
`return $this->redirect($this->referrer('/default'));` when the app is in a
subfolder as the base get appended twice.

### CODE REVIEW

**Comment by markstory** [comment]
Tests [are failing in relevant ways](https://travis-ci.org/cakephp/cakephp/jobs/124108056).

File: tests/TestCase/Controller/ControllerTest.php [hunk 1]
```diff
@@ -626,7 +626,7 @@ public function testRefererSlash()
 
         $controller = new Controller($request);
         $result = $controller->referer('/some/path', true);
-        $this->assertEquals('/base/some/path', $result);
+        $this->assertEquals('/some/path', $result);
```
Comment (markstory) [inline tests/TestCase/Controller/ControllerTest.php]: This looks like it could end up breaking applications that are relying on this.

Comment (ADmad) [inline tests/TestCase/Controller/ControllerTest.php]: This issue only occurs when running the app in a subfolder which would be pretty rare in production. Plus I believe the major usage of getting referrer is to use it for redirection so anyone having their app in subfolder would be removing removing the base himself for the redirection to work properly. So they probability of affecting existing app is very low. That said I am fine with moving the patch to 3.next.

**Comment by ADmad** [comment]
Code updated. There shouldn't be any test failures anymore.

File: tests/TestCase/Controller/ControllerTest.php (hunk 1 above)
Comment (markstory) [inline tests/TestCase/Controller/ControllerTest.php]: > This issue only occurs when running the app in a subfolder which would be pretty rare in production.

I'm not so confident on that.

Comment (dereuromark) [inline tests/TestCase/Controller/ControllerTest.php]: People never should use that kind of setup IMO ;) But we all know they do.
Is there a way to maybe provide this BC - so you opt in for the "new correct" way?
Maybe with some Configure key then? Or we can apply this for the next 3.3 version with a migration note.

**Comment by ADmad** [comment]
I am not favour of adding a Configure key for something like this. I can reopen against 3.next if that's the consensus else the PR can just be closed.

**Comment by markstory** [comment]
I think moving to 3.next is a good way to correct the behaviour and not surprise people quite as easily.

**Pull request closed without merging**